File: 2025/day1.2.py
def main():
    dial = 50
    password = 0

    with open('day1.txt', 'r') as f:
        for line in f:
            line = line.strip()
            if not line:
                continue

            # Track if dial is 0 at the beginning of the loop
            was_on_0 = (dial == 0)

            direction = line[0]
            clicks = int(line[1:])

            # Add integer division of clicks by 100 to password
            password += clicks // 100

            clicks = clicks % 100

            if direction == 'L':
                dial -= clicks
            elif direction == 'R':
                dial += clicks
            else:
                print(f"Error: Invalid direction in line: {line}")
                return

            # If dial is less than 0, increase password by 1 but only if was_on_0 is false
            if dial < 0:
                if not was_on_0:
                    password += 1
                dial += 100

            # If dial is bigger than or equals 100, increase password by 1 but only if dial is not 100
            if dial >= 100:
                if dial != 100:
                    password += 1
                dial -= 100

            if dial == 0:
                password += 1

    print(password)

if __name__ == '__main__':
    main()

File: 2025.go/day1.2.go
//go:build ignore

package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
)

func main() {
	file, err := os.Open("../2025/day1.txt")
	if err != nil {
		log.Fatalf("Failed to open file: %v", err)
	}
	defer file.Close()

	dial := 50
	password := 0

	scanner := bufio.NewScanner(file)
	lineNum := 0
	for scanner.Scan() {
		lineNum++
		line := scanner.Text()

		if len(line) < 2 {
			log.Fatalf("Line %d is too short: %s", lineNum, line)
		}

		direction := line[0]
		clicksStr := line[1:]

		clicks, err := strconv.Atoi(clicksStr)
		if err != nil {
			log.Fatalf("Failed to parse clicks on line %d: %s", lineNum, line)
		}

		// Track if dial is 0 at the beginning of the loop
		wasOn0 := dial == 0

		// Integer division of clicks by 100, add to password
		password += clicks / 100

		// Mod clicks by 100
		clicks = clicks % 100

		// Update dial based on direction
		switch direction {
		case 'L':
			dial -= clicks
		case 'R':
			dial += clicks
		default:
			log.Fatalf("Invalid direction on line %d: %s", lineNum, line)
		}

		// Handle negative dial
		if dial < 0 {
			if !wasOn0 {
				password++
			}
			dial += 100
		}

		// Handle dial >= 100
		if dial >= 100 {
			if dial != 100 {
				password++
			}
			dial -= 100
		}

		// Check if dial is 0
		if dial == 0 {
			password++
		}
	}

	if err := scanner.Err(); err != nil {
		log.Fatalf("Error reading file: %v", err)
	}

	fmt.Println(password)
}
